src/provider/p_ibmca.c

```diff
@@ -0,0 +1,1392 @@
+    return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_BUILDINFO);
+    if (p != NULL && !OSSL_PARAM_set_utf8_ptr(p, IBMCA_VERSION)) {
+        put_error_ctx(provctx, IBMCA_ERR_INTERNAL_ERROR,
+                      "OSSL_PARAM_set_utf8_ptr failed");
+        return 0;
+    }
+    p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_STATUS);
+    if (p != NULL && !OSSL_PARAM_set_int(p, 1)) {
+        put_error_ctx(provctx, IBMCA_ERR_INTERNAL_ERROR,
+                      "OSSL_PARAM_set_int failed");
+        return 0;
+    }
+
+    return 1;
+}
+
+static const OSSL_ALGORITHM *ibmca_query(void *vprovctx, int operation_id,
+                                         int *no_cache)
+{
+    const struct ibmca_prov_ctx *provctx = vprovctx;
+
+    if (provctx == NULL)
+        return NULL;
+
+    *no_cache = 0;
+
+    ibmca_debug_ctx(provctx, "provctx: %p operation_id: %d", provctx,
+                    operation_id);
+
+    if (operation_id < 0 || operation_id > OSSL_OP__HIGHEST)
+        return NULL;
+
+    ibmca_debug_ctx(provctx, "algorithms: %p",
+                    provctx->algorithms[operation_id]);
+
+    return provctx->algorithms[operation_id];
+}
+
+static const OSSL_ITEM *ibmca_get_reason_strings(void *vprovctx)
+{
+    const struct ibmca_prov_ctx *provctx = vprovctx;
+
+    ibmca_debug_ctx(provctx, "provctx: %p", provctx);
+    return ibmca_reason_strings;
+}
+
+static int ibmca_prov_get_capabilities(void *vprovctx, const char *capability,
+                                       OSSL_CALLBACK *cb, void *arg)
+{
+    const struct ibmca_prov_ctx *provctx = vprovctx;
+    int i, k, rc = 0;
+    const struct ibmca_mech_capability *cap;
+
+    if (capability == NULL || cb == NULL)
+        return 0;
+
+    ibmca_debug_ctx(provctx, "provctx: %p capability: %s", provctx,
+                    capability);
+
+    for (i = 0; ica_mech_infos[i].algo != NULL; i++) {
+        for (cap = ica_mech_infos[i].capabilities;
+             cap != NULL && cap->capability != NULL; cap++) {
+            if (strcasecmp(capability, cap->capability) != 0)
+                continue;
+
+            ibmca_debug_ctx(provctx, "algorithm '%s' supports this capability",
+                            ica_mech_infos[i].algo);
+
+            for (k = 0; cap->details[k] != NULL; k++) {
+                rc = cb(cap->details[k], arg);
+                if (rc == 0)
+                    break;
+            }
+
+            if (rc == 0)
+                break;
+        }
+    }
+
+    return rc;
+}
+
+static int ibmca_prov_self_tests(void *vprovctx)
+{
+    const struct ibmca_prov_ctx *provctx = vprovctx;
+
+    ibmca_debug_ctx(provctx, "provctx: %p", provctx);
+
+    if (provctx->fips == 0)
+        return 1;
+
+    /*
+     * libica performs FIPS self-tests if running in FIPS mode. If any of
+     * the fips tests failed, additional flags are on.
+     */
+    return (provctx->ica_fips_status == ICA_FIPS_MODE) ? 1 : 0;
```

Comment:
Right, this is certainly not enough to claim FIPS compliance. May thinking was that we start with this minimal FIPS support, any add the full FIPS support including own selftests and HMAC checks in a later stage. Before we spend the effort for this I would like to know what the distros think about full FIPS support of the provider.

However, it is of course questionable if setting `fips=yes` for the property definitions of the IBMCA provider is correct with this minimal FIPS support.....

How about declaring this minimum FIPS support as an experimental 'feature' and making it compile-time dependent on `#ifdef IBMCA_FIPS` which can be defined at configure time ? That way we keep the code, but it won't be active by default. Once we add full FIPS support at a later time, we can easily enable it.

This would of course also affect other places, like the fips config option, man pages, documentation, etc. where the FIPS mode must be removed.

Reply:
Personally I do not like build options.  Lets go the even simpler route and keep the code, but document in the man page that it is experimental and currently not FIPS certified.  I guess this will be enough and not introduce ugliness in the form of `#ifdef`s in the code.

Reply:
ok fine with me.

How about the following note in the man page at the end of the `fips = yes | no` configuration setting description:

_Note: The IBMCA provider is currently not FIPS certified. It does not itself perform any FIPS self tests nor  an  integrity check which would be required to be FIPS certifiable. It is only ensured that libica-cex library has successfully performed its self tests and integrity checks when FIPS mode is enabled.  If you do not  want  that  the  IBMCA provider  is used for property queries that include fips=yes, then disable FIPS mode of the IBMCA provider by setting fips = no in the provider configuration._